llvm/include/llvm/Support/Progress.h

```diff
@@ -0,0 +1,256 @@
+//===-- Progress - Generic infrastructure to report progress ----*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_SUPPORT_PROGRESS_H
+#define LLVM_SUPPORT_PROGRESS_H
+
+extern "C" {
+#include <unistd.h>
+}
+
+#include <algorithm>
+#include <chrono>
+#include <optional>
+#include <set>
+#include <vector>
+
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/Twine.h"
+#include "llvm/Support/ManagedStatic.h"
+#include "llvm/Support/RWMutex.h"
+
+namespace llvm {
+
+class Progress;
+class TaskStack;
+
+/// Track advancement of a task
+///
+/// A task can have a name and optionally a set of maximum steps.
+class Task {
+protected:
+  TaskStack &Stack;
+  int64_t StepIndex = -1;
+  std::optional<int64_t> TotalSteps;
+
+private:
+  std::string TaskName;
+  std::string StepName;
+  bool SubtaskCreated = false;
+  bool Completed = false;
+  bool CurrentStepHasSingleSubtask = false;
+
+public:
+  Task(std::optional<int64_t> TotalSteps, const llvm::Twine &TaskName = "");
+
+  Task(TaskStack &Stack, std::optional<int64_t> TotalSteps,
+       const llvm::Twine &TaskName = "");
+
+  Task(const Task &) = delete;
+  Task(Task &&) = delete;
+
+  ~Task();
+
+public:
+  llvm::StringRef name() const { return TaskName; }
+  llvm::StringRef stepName() const { return StepName; }
+  int64_t stepIndex() const {
+    return std::max(static_cast<int64_t>(0), StepIndex);
+  }
+  std::optional<int64_t> totalSteps() const { return TotalSteps; }
+  bool subtaskCreated() const { return SubtaskCreated; }
+  bool currentStepHasSingleSubtask() const {
+    return CurrentStepHasSingleSubtask;
+  }
+  bool completed() const { return Completed; }
+  size_t index() const;
+
+  TaskStack &stack() { return Stack; }
+  const TaskStack &stack() const { return Stack; }
+
+public:
+  /// \param NewStepName the name of the new step starting after this advance
+  ///        invocation.
+  /// \param SingleSubtask whether we can expect 0 or 1 subtask being created
+  ///        during this step. If true, will assert in case multiple subtasks
+  ///        are create during this step.
+  void advance(const llvm::Twine &NewStepName = "", bool SingleSubtask = false);
+
+  void setSubtaskCreated();
+
+  void complete();
+
+};
+
+/// Represents the set of tasks started but not completed in an execution thread
```

Comment:
It seems to me that `Task` is always expected to have an associated parent `TaskStack`. At this point why not just make `TaskStack` own `Task`s?

Reply:
Because I like to declare `Task`s as local variables.
Can you suggest a solution as user-friendly where `TaskStack` owns `Task`?

Reply:
Ah RAII is nice indeed.

One solution could be to have `TaskStack` own `Task`s and `TaskStack.pushNewTask(Size, Name)` return a thin RAII object that holds a reference to `TaskStack` and that just calls `TaskStack.back().complete(); TaskStack.unregisterTask(TaskStack.back());` on destruction.

But it's not very much different than what you have now.

Reply:
Well, the first syntax requires you to understand `Task` which is a RAII object basically. So the kind of logic you have to understand is the same.

But yes, the first syntax is much more concise, and the second does not really other any advantage compared to it, except being more explicit for a newcomer that is curious about what happens under the hood.

So yeah, let's stick with the first syntax that makes things ergonomics for everyday use.